src/backend/cdb/cdbpath.c

```diff
@@ -1109,19 +1140,30 @@ cdbpath_motion_for_join(PlannerInfo *root,
 
 			if (CdbPathLocus_IsReplicated(other->locus))
 			{
-				Assert(root->upd_del_replicated_table > 0);
-
 				/*
-				 * It only appear when we UPDATE a replicated table.
-				 * All the segment which replicated table storaged must execute
-				 * the plan to delete tuple on himself, so if the segments count
-				 * of broadcast(locus is Replicated) if less than the replicated
-				 * table, we can not execute the plan correctly.
+				 * In case when we UPDATE/DELETE a replicated table,
+				 * upd_del_replicated is set to true. Here we need to be sure
+				 * that all the segments, at which replicated table are
+				 * storaged, must execute the plan to delete tuple on himself,
+				 * so if the segments count of broadcast(locus is Replicated)
+				 * is less than the replicated table, we can not execute the
+				 * plan correctly.
 				 *
 				 * TODO:Can we modify(or add) the broadcast motion for this case?
+				 *
+				 * Second possible case: DML over replicated table inside CTE
+				 * with SELECT above. When we join the result of modifying CTE
+				 * with another table, which is too replicated, but has less
+				 * numsegments, then we want DML part of query be executed on
+				 * its all, and SELECT part of query be executed on less
+				 * numsegments. "Explicit Gather Motion" will use less
+				 * numsegments in execMotionSender(), so our result will not
+				 * be flacky because of new segments without a data.
 				 */
-				Assert(CdbPathLocus_NumSegments(segGeneral->locus) <=
-					   CdbPathLocus_NumSegments(other->locus));
+				Assert((root->upd_del_replicated_table > 0 &&
+					   CdbPathLocus_NumSegments(segGeneral->locus) <=
+					   CdbPathLocus_NumSegments(other->locus)) ||
+					   hasModifyingCTESelectRecur(root));
```

Comment:
Can you explain difference in handling UPDATE/DELETE with and without wrapping CTE? Code before patch assumes that it's possible to face join between Replicated and SegmentGeneral distribution. And the only expected case is number of segments of SegmentGeneral is less or equal of number of segments of Replicated. Why do we ready to ignore this strange limitation inside CTE?

Reply:
This limitation was originally aimed at handling the case of common UPDATE/DELETE query. When doing UPDATE ... FROM and updating specific rows on join condition, one need to be sure that DML operation will perform on all the segments with data. This `<=` constraint was introduced in this [commit](https://github.com/greenplum-db/gpdb/commit/b32d6110d9a9642c9f675525b7312e3b927b377b) and is used for the situation when replicated table with SegmentGeneral locus is updated on join condition with other relation (locus Replicated). And such restriction means that one cannot update SegmentGeneral table by join with table, which is distiributed on less number of segments than the table one want to update. 
For our situation with modifying CTE this limitaion is unnecessary, because we just want to perform common join with 2 relations: replicated return of CTE (locus Replicated) and replciated table (SegmentGeneral). And here it is appropriate to `CdbPathLocus_NumSegments(segGeneral->locus)` be `>`  `CdbPathLocus_NumSegments(other->locus)`

Reply:
Comments and modification strict below contradicts with your statement too. It's useless and potentially danger to modify numsegments for replicated distribution generated by DML operation.